lib/Conversion/TritonToStructured/TritonToStructuredPass.cpp

```diff
@@ -41,13 +57,244 @@ class TritonToStructuredPass
                 tts::TritonStructuredDialect>();
   }
 
-  void runOnOperation() override {
+  LogicalResult convertToPointerTupleWithOffsetsAndStrides() {
     auto moduleOp = getOperation();
 
+    RewritePatternSet patterns(&getContext());
+
+    auto context = &getContext();
+    OneToNTypeConverter converter;
+    converter.addConversion([](Type type) { return type; });
+
+    // We are doing a 1->1 type conversion here, where a triton pointer type
+    // maps to a tuple of {pointer, offset_0, offset_1,..., stride_0,
+    // stride_1,...} type.
+    //
+    // Case 1: Unstructured pointers (tensor<!tt.ptr<type>>)
+    converter.addConversion([context](RankedTensorType tensorType,
+                                      SmallVectorImpl<Type> &types)
+                                -> std::optional<LogicalResult> {
+      if (auto ptrType =
+              dyn_cast<triton::PointerType>(tensorType.getElementType())) {
+        SmallVector<Type> tupleTypes{tensorType};
+        // Each tensor of rank k gets k values for its offsets and k values for
+        // its strides, all of which has Index type.
+        tupleTypes.append(tensorType.getRank() * 2, IndexType::get(context));
+        types = SmallVector<Type>{TupleType::get(context, tupleTypes)};
+        return success();
+      }
+      return std::nullopt;
```

Comment:
nit, `return failure();`

Reply:
This is actually intended and caught me off guard too. There's a subtle difference between returning `failure()` and `std::nullopt`. From the [documentation](https://github.com/llvm/llvm-project/blob/3210bce90f8678d52ea67f812bfe81d1c9de7597/mlir/include/mlir/Transforms/DialectConversion.h#L152):

> If `std::nullopt` is returned, the converter is allowed to try another conversion function to perform the conversion.

Say we have type `tensor<4x256xbf16>` which is a `RankedTensorType`. Even though this `RankedTensorType` matches the converter that handles the tuple conversion, we want to keep this type as is because the inner type isn't a pointer. So by returning `failure()`, the TypeConverters will stop trying the remaining converters. In our case, the last type converter which simply returns the same type is skipped. And because the conversion for this type has failed, the whole conversion process is also skipped.

Relevant links to the implementation:
+ https://github.com/llvm/llvm-project/blob/cb5dc1faa8b3702e0d03426ee5dfc5e1b903ec47/mlir/lib/Transforms/Utils/DialectConversion.cpp#L2958
+ https://github.com/llvm/llvm-project/blob/cb5dc1faa8b3702e0d03426ee5dfc5e1b903ec47/mlir/lib/Transforms/Utils/DialectConversion.cpp#L3033

Reply:
worth a comment in the code!